contrib/parseq-batching/src/main/java/com/linkedin/parseq/batching/BatchingStrategy.java

```diff
@@ -0,0 +1,282 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+
+import com.linkedin.parseq.Task;
+import com.linkedin.parseq.batching.BatchImpl.BatchBuilder;
+import com.linkedin.parseq.batching.BatchImpl.BatchEntry;
+import com.linkedin.parseq.internal.ContextImpl;
+import com.linkedin.parseq.internal.PlanContext;
+import com.linkedin.parseq.promise.CountDownPromiseListener;
+import com.linkedin.parseq.promise.PromiseListener;
+import com.linkedin.parseq.promise.Promises;
+import com.linkedin.parseq.promise.SettablePromise;
+import com.linkedin.parseq.trace.Relationship;
+import com.linkedin.parseq.trace.ShallowTraceBuilder;
+import com.linkedin.parseq.trace.TraceBuilder;
+
+/**
+ * {@code BatchingStrategy} helps building "batching clients" in ParSeq. "Client" means on object that given {@code K key}
+ * provides a task that returns {@code T value}. "Batching" means that it can group together keys to resolve values
+ * in batches. The benefit of this approach is that batching happens transparently in the background and user's code
+ * does not have to deal with logic needed to implement batching.
+ * <p>
+ * Example of a batching client might be ParSeq client for a key-value store that provides batch get operation. For
+ * the sake of simplicity of the example we are using dummy, synchronous key-value store interface:
+ * <blockquote><pre>
+ *  interface KVStore {
+ *    String get(Long key);
+ *    Map{@code <Long, String>} batchGet(Collection{@code <Long>} keys);
+ *  }
+ * </pre></blockquote>
+ *
+ * We can then implement a {@code BatchingStrategy} in the following way:
+ * <blockquote><pre>
+ *  public static class BatchingKVStoreClient extends BatchingStrategy{@code <Integer, Long, String>} {
+ *    private final KVStore _store;
+ *    public BatchingKVStoreClient(KVStore store) {
+ *      _store = store;
+ *    }
+ *
+ *    {@code @Override}
+ *    public void executeBatch(Integer group, Batch{@code <Long, String>} batch) {
+ *      Map{@code <Long, String>} batchResult = _store.batchGet(batch.keys());
+ *      batch.foreach((key, promise) {@code ->} promise.done(batchResult.get(key)));
+ *    }
+ *
+ *    {@code @Override}
+ *    public void executeSingleton(Integer group, Long key, BatchEntry{@code <String>} entry) {
+ *      entry.getPromise().done(_store.get(key));
+ *    }
+ *
+ *    {@code @Override}
+ *    public Integer classify(Long entry) {
+ *      return 0;
+ *    }
+ *  }
+ * </pre></blockquote>
+ *
+ * In above example there is an assumption that all keys can be grouped together. This is why method {@code classify()}
+ * trivially returns a constant {@code 0}. In practice {@code classify()} returns an equivalence class. All keys that
+ * returns equal equivalence class will constitute a batch.
```

Comment:
I think it would be clearer to say "Only keys that have the same group will be batched together"